feat(data): Storage and GraphQL API authorization guidance; add additional guidance for error handling and real-time data access; remove existing AppSync SDK docs; update docs comments (#5545)
aws-amplify/docs: src/fragments/lib/graphqlapi/js/working-with-files.mdx

</Block>
</BlockSwitcher>

> ## Working with real-time data

Reviewer: Seems mis-titled.

This section casually suggests realtime data as a solution to further minimize the effects of data sync issues on clients. But, the section is more generally about "data consistency" and/or "defensive programming".

Author: Will update

> ## Working with real-time data

Reviewer 2: ```suggestion
## Working with real-time data and files
```

> The recommended access patterns in these docs minimize the occurrence of records that point to non-existent and/or orphaned files. This optimizes for read latency by ensuring clients *rarely* attempt to fetch a non-existent file from Storage. However, any app that deletes files can inherently cause records *on-device* to point to non-existent files.

Reviewer: > The recommended access patterns in these docs minimize the occurrence of records that point to non-existent and/or orphaned files.

To my eyes, this sound like our patterns are trying to "minimize the occurrence of records that point to orphaned files". And, that's just nonsensical. Our readers will figure it out, but I think the phrasing is just awkward.

On top of that, while our patterns *do* attempt to remove files instead of orphaning them, they explicitly favor creating orphans over creating mirages. If you want to mention both, I might state it more like that.

```suggestion
The recommended access patterns in these docs attempt remove deleted files, but favor leaving orphans over leaving records that point to non-existent files. This optimizes for read latency by ensuring clients *rarely* attempt to fetch a non-existent file from Storage. However, any app that deletes files can inherently cause records *on-device* to point to non-existent files.
```

One example is when we [create an API record, associate the Storage file with that record, and then retrieve the file's signed URL](#create-a-record-with-an-associated-file). "Device A" calls the GraphQL API to create `API_Record_1`, and then associates that record with `First_Photo`. Before "Device A" is about to retrieve the signed URL, "Device B" might query `API_Record_1`, delete `First_Photo`, and update the record accordingly. However, "Device A" is still using the old `API_Record_1`, which is now out-of-date. Even though the shared global state is correctly in sync at every stage, the individual device ("Device A") has an out-of-date record that points to a non-existent file. Similar issues can conceivably occur for updates. Depending on your app, some of these mismatches can be minimized *even more* with [real-time data / GraphQL subscriptions](https://docs.amplify.aws/lib/graphqlapi/subscribe-data/).

> It is important to understand when these mismatches can occur in your app, and to add meaningful error handling around these edges. For the purposes of these docs, this guide does not include exhaustive error handling, real-time subscriptions, re-querying of outdated records, or attempts to retry failed operations. However, these are all important considerations for a production-level application.

Reviewer 2: ```suggestion
It is important to understand when these mismatches can occur and to add meaningful error handling around these cases. This guide does not include exhaustive error handling, real-time subscriptions, re-querying of outdated records, or attempts to retry failed operations. However, these are all important considerations for a production-level application.
```

Reviewer 2: Edited for succinctness
